### ISSUE TITLE
Multiple benchmarks panic due to bounds check

### ISSUE THREAD

**Issue by Antelope13** [slide-rs/specs#200]
When running `cargo bench` I get panics in `create_now`, `create_now_with_storage`, and `maintain_add_later`. Each one seems to have ~50% chance to panic. Occasionally all benches will pass. Tested it on all combinations of stable/nightly/msvc/gnu on a Windows 7 and Windows 10 machine. 

`cargo test` passes all tests.

`create_now` and `create_now_with_storage` both panic at `specs::world::Allocator::allocate`
> test create_now              ... thread 'main' panicked at 'index out of bounds: the len is 4096 but the index is 4096',
>  C:\projects\rust\src\liballoc\vec.rs:1552
> stack backtrace:
>    0: std::sys_common::backtrace::_print
>    1: std::panicking::default_hook::{{closure}}
>    2: std::panicking::default_hook
>    3: std::panicking::rust_panic_with_hook
>    4: std::panicking::begin_panic
>    5: std::panicking::begin_panic_fmt
>    6: rust_begin_unwind
>    7: core::panicking::panic_fmt
>    8: core::panicking::panic_bounds_check
>    9: specs::world::Allocator::allocate
>   10: specs::world::World::create_entity
>   11: world::create_now
>   12: test::run_test
>   13: test::run_tests_console
>   14: test::test_main
>   15: test::test_main_static
>   16: _rust_maybe_catch_panic
>   17: std::rt::lang_start
>   18: _tmainCRTStartup
>   19: mainCRTStartup
>   20: unit_addrs_search
> error: bench failed

`maintain_add_later` panics at `specs::world::Allocator::allocate_atomic`
> test maintain_add_later      ... thread 'main' panicked at 'index out of bounds: the len is 4096 but the index is 4096',
>  C:\projects\rust\src\liballoc\vec.rs:1552
> stack backtrace:
>    0: std::sys_common::backtrace::_print
>    1: std::panicking::default_hook::{{closure}}
>    2: std::panicking::default_hook
>    3: std::panicking::rust_panic_with_hook
>    4: std::panicking::begin_panic
>    5: std::panicking::begin_panic_fmt
>    6: rust_begin_unwind
>    7: core::panicking::panic_fmt
>    8: core::panicking::panic_bounds_check
>    9: specs::world::Allocator::allocate_atomic
>   10: test::ns_iter_inner
>   11: world::maintain_add_later
>   12: test::run_test
>   13: test::run_tests_console
>   14: test::test_main
>   15: test::test_main_static
>   16: _rust_maybe_catch_panic
>   17: std::rt::lang_start
>   18: _tmainCRTStartup
>   19: mainCRTStartup
>   20: unit_addrs_search
> error: bench failed

When it does pass:
> running 12 tests
> test create_now              ... bench:          29 ns/iter (+/- 3)
> test create_now_with_storage ... bench:          73 ns/iter (+/- 14)
> test create_pure             ... bench:          46 ns/iter (+/- 8)
> test delete_later            ... bench:           0 ns/iter (+/- 0)
> test delete_now              ... bench:           0 ns/iter (+/- 0)
> test delete_now_with_storage ... bench:           0 ns/iter (+/- 0)
> test join_multi_threaded     ... bench:      38,379 ns/iter (+/- 2,427)
> test join_single_threaded    ... bench:     170,115 ns/iter (+/- 7,501)
> test maintain_add_later      ... bench:         138 ns/iter (+/- 30)
> test maintain_delete_later   ... bench:          40 ns/iter (+/- 2)
> test maintain_noop           ... bench:          43 ns/iter (+/- 7)
> test world_build             ... bench:      10,248 ns/iter (+/- 315)

**Comment by torkleyy** [comment]
Hi @Antelope13!

That's correct, this is a known issue. Unfortunately, there doesn't seem to be an obvious solution.

The main problem is that the current benchmarking in Rust is pretty unstable and should be replaced, but nobody has found the time to do so yet.

**Comment by Aceeri** [comment]
Does this still happen? I think this was fixed in a later nightly build.

**Comment by Antelope13** [comment]
I can confirm that this is still an open issue. It looks like it occurs when more than 16_777_216 (2^24) entities are created. I think it's a limitation of [hibitset](https://github.com/slide-rs/hibitset)? Is there any way to increase this limit?

It's an absurd number of entities for a game but I've been experimenting with using specs as a general-purpose simulation engine and I've bumped into this limit more than once.

**Comment by torkleyy** [comment]
This limitation stems from the fact that each layer of the bitset has one bit representing 8 bits in the underlying layer. So we have got up to 64 = 2^6 bits in the upmost layer. => 2^6 * 2^6 * 2^6 * 2^6 = 2^24.

TL;DR: If we want to increase that limit we need to add more layers.

**Comment by WaDelma** [comment]
If only we had integer generics and we could give the user a choice on it...

**Comment by torkleyy** [comment]
Well we can easily emulate integer generics here, but the bigger problem would be how you want things to work with one type (`Entities`) then.

**Comment by torkleyy** [comment]
@Antelope13 So the only reasonable way I see for bumping the limit would be using a `--cfg` flag (we can't use features since they're only for additive things). Making everything generic over the limit wouldn't be very nice in my opinion.

**Comment by torkleyy** [comment]
I have a solution for the benchmarking problem. I started working on a branch using [criterion.rs](https://github.com/japaric/criterion.rs) for the benchmarking: https://github.com/torkleyy/specs/tree/bench

That allows us to use a setup function in order to clear state between iterations or alternatively we can set a maximum number of iterations.

**Comment by jonas-schievink** [comment]
Making a higher limit a feature instead of a cfg should be fine, it has no impact (except a slight performance regression) on code that's not using the feature

**Comment by torkleyy** [comment]
I think I'll just make the `World` generic over the allocator, providing a default for the current one.
